com.microsoft.mrtk.spatialmanipulation/ObjectManipulator/ObjectManipulator.cs

```diff
@@ -570,106 +639,158 @@ public override void ProcessInteractable(XRInteractionUpdateOrder.UpdatePhase up
+
+            // implement critically dampened spring force, scaled to mass-independent frequency
+            float omega = Mathf.PI / springForceSoftness;  // angular frequency, sqrt(k/m)
+
+            Vector3 distance = HostTransform.position - targetTransform.Position;
 
-                    if (axis.IsValidVector())
+            // when player is moving, we need to anticipate where the targetTransform is going to be one time step from now
+            distance -= referenceFrameVelocity * Time.fixedDeltaTime;
+
+            var velocity = rigidBody.velocity;
+
+            var acceleration = -distance * omega * omega;  // acceleration caused by spring force
+
+            var accelerationMagnitude = acceleration.magnitude;
+
+            // apply springForceLimit only for slow-moving body (e.g. pressed against wall)
+            // when body is already moving fast, also allow strong acceleration
+            var maxAcceleration = Mathf.Max(springForceLimit, 10 * velocity.magnitude / Time.fixedDeltaTime);
+
+            if (accelerationMagnitude > maxAcceleration)
+            {
+                acceleration *= maxAcceleration / accelerationMagnitude;
+            }
+
+            // Apply damping - mathematically, we need e^(-2 * omega * dt)
+            // To compensate for the finite time step, this is split in two equal factors,
+            // one applied before, the other after the spring force
+            // equivalent with applying damping as well as spring force continuously
+            float halfDampingFactor = Mathf.Exp(-springForceDamping * omega * Time.fixedDeltaTime);
+
+            velocity -= referenceFrameVelocity;  // change to the player's frame of reference before damping
+
+            velocity *= halfDampingFactor;  // 1/2 damping
+            velocity += acceleration * Time.fixedDeltaTime; // integration step of spring force
+            velocity *= halfDampingFactor;  // 1/2 damping
+
+            velocity += referenceFrameVelocity;  // change back to global frame of reference
+
+            rigidBody.velocity = velocity;
+
+            if (applyTorque)
+            {
+                var relativeRotation = targetTransform.Rotation * Quaternion.Inverse(HostTransform.rotation);
+                relativeRotation.ToAngleAxis(out float angle, out Vector3 axis);
+
+                if (axis.IsValidVector())
+                {
+                    if (angle > 180f)
                     {
-                        if (angle > 180f)
-                        {
-                            angle -= 360f;
-                        }
-                        rigidBody.angularVelocity = ((1f - Mathf.Pow(rotateLerpTime, Time.deltaTime)) / Time.deltaTime) * (angle * Mathf.Deg2Rad * axis.normalized);
+                        angle -= 360f;
                     }
+                    rigidBody.angularVelocity = ((1f - Mathf.Pow(rotateLerpTime, Time.fixedDeltaTime)) / Time.fixedDeltaTime) * (angle * Mathf.Deg2Rad * axis.normalized);
```

Comment:
Is this method only expected to be called from `FixedUpdate`? I guess, either way, you _could_ leave this as `Time.deltaTime` if you wanted, since that returns `Time.fixedDeltaTime` automatically when called from `FixedUpdate`.

https://docs.unity3d.com/ScriptReference/Time-deltaTime.html

> When this is called from inside [MonoBehaviour.FixedUpdate](https://docs.unity3d.com/ScriptReference/MonoBehaviour.FixedUpdate.html), it returns [Time.fixedDeltaTime](https://docs.unity3d.com/ScriptReference/Time-fixedDeltaTime.html).

Reply:
Interesting, indeed. I never noticed that sentence in the documentation.

In any case, it does not really matter which one to use here: The function is called only from FixedUpdate.

Reply:
Works for me!